### PR TITLE
{AKS} feat: ga update for azure monitor for metrics addon (managed prometheus)

### PR DESCRIPTION
**Pull request by bragi92**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

```az aks update -n kaveeshcli22 -g kaveeshcli --enable-azuremonitormetrics --enable-windows-recording-rules```

```az aks create -n kaveeshcli22 -g kaveeshcli --location westeurope --enable-azuremonitormetrics --enable-windows-recording-rules --azure-monitor-workspace-resource-id "{full_id}" --grafana-resource-id "{full_id}"```

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

**Description**

GA update for Azure Monitor Metrics. The changes for public preview where merged into the az cli extensions repo in the following PR: https://github.com/Azure/azure-cli-extensions/pull/5428

The GA update PR for the azure cli repo is here : https://github.com/Azure/azure-cli/pull/26201

**Testing Guide**

```az aks create``` and ```az aks update``` is being updated with the following parameters:

  - name: --enable-azuremonitormetrics
    type: bool
    short-summary: Enable Azure Monitor Metrics Profile
  - name: --azure-monitor-workspace-resource-id
    type: string
    short-summary: Resource ID of the Azure Monitor Workspace
  - name: --ksm-metric-labels-allow-list
    type: string
    short-summary: Comma-separated list of additional Kubernetes label keys that will be used in the resource' labels metric. By default the metric contains only name and namespace labels. To include additional labels provide a list of resource names in their plural form and Kubernetes label keys you would like to allow for them (e.g. '=namespaces=[k8s-label-1,k8s-label-n,...],pods=[app],...)'. A single '*' can be provided per resource instead to allow any labels, but that has severe performance implications (e.g. '=pods=[*]').
  - name: --ksm-metric-annotations-allow-list
    type: string
    short-summary: Comma-separated list of additional Kubernetes label keys that will be used in the resource' labels metric. By default the metric contains only name and namespace labels. To include additional labels provide a list of resource names in their plural form and Kubernetes label keys you would like to allow for them (e.g.'=namespaces=[k8s-label-1,k8s-label-n,...],pods=[app],...)'. A single '*' can be provided per resource instead to allow any labels, but that has severe performance implications (e.g. '=pods=[*]').
  - name: --grafana-resource-id
    type: string
    short-summary: Resource ID of the Azure Managed Grafana Workspace
  - name: --enable-windows-recording-rules
    type: bool
    short-summary: Enable Windows Recording Rules when enabling the Azure Monitor Metrics addon
  - name: --disable-azuremonitormetrics
    type: bool
    short-summary: Disable Azure Monitor Metrics Profile. This will delete all DCRA's associated with the cluster, any linked DCRs with the data stream = prometheus-stream and the recording rule groups created by the addon for this AKS cluster.

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/aks-preview/azext_aks_preview/custom.py [hunk 1]
```diff
@@ -608,8 +618,38 @@ def aks_create(
     except DecoratorEarlyExitException:
         # exit gracefully
         return None
-    # send request to create a real managed cluster
-    return aks_create_decorator.create_mc(mc)
+
+    try:
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/custom.py:622]: should use the create method from the decorator

**Review by FumingZhang — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/custom.py (hunk 1 above)
Comment (bragi92) [reply src/aks-preview/azext_aks_preview/custom.py]: updated, thanks for the feedback!

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/amg/link.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    # Setting up AMW Integration
+    targetGrafanaArmPayload = grafanaArmResponse.json()
+    if targetGrafanaArmPayload["properties"] is None:
+        raise CLIError("Invalid grafana payload to add AMW integration")
+    if "grafanaIntegrations" not in json.dumps(targetGrafanaArmPayload):
+        targetGrafanaArmPayload["properties"]["grafanaIntegrations"] = {}
+    if "azureMonitorWorkspaceIntegrations" not in json.dumps(targetGrafanaArmPayload):
+        targetGrafanaArmPayload["properties"]["grafanaIntegrations"]["azureMonitorWorkspaceIntegrations"] = []
+    amwIntegrations = targetGrafanaArmPayload["properties"]["grafanaIntegrations"]["azureMonitorWorkspaceIntegrations"]
+    if amwIntegrations != [] and azure_monitor_workspace_resource_id in json.dumps(amwIntegrations).lower():
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/amg/link.py:66]:
```suggestion
    if amwIntegrations != [] and azure_monitor_workspace_resource_id in json.dumps(amwIntegrations).lower():
```
```suggestion
    if amwIntegrations and azure_monitor_workspace_resource_id in json.dumps(amwIntegrations).lower():
```

**Review by FumingZhang — COMMENTED** [review]
Queued [live test pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=73045802&view=logs&j=b162b355-d59d-5864-ce0f-0a70f12dd28b&t=a26fe913-e2be-5062-b08e-d15f1acc7ea2) to verify the change, test passed

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/amw/defaults.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import json
+from azext_aks_preview.azuremonitormetrics.deaults import get_default_region
+from azext_aks_preview.azuremonitormetrics.responseparsers.amwlocationresponseparser import parseResourceProviderResponseForLocations
+from azext_aks_preview.azuremonitormetrics.constants import RP_LOCATION_API
+from knack.util import CLIError
+
+
+def get_supported_rp_locations(cmd, rp_name):
+    from azure.cli.core.util import send_raw_request
+    supported_locations = []
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/amw/defaults.py:14]:
```suggestion
```

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/amw/helper.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        resource = resources.get_by_id(
+            azure_monitor_workspace_resource_id, MAC_API)
+        return resource.location.lower()
+    except HttpResponseError as ex:
+        raise ex
+
+
+def get_azure_monitor_workspace_resource(cmd, cluster_subscription, cluster_region, raw_parameters):
+    azure_monitor_workspace_resource_id = raw_parameters.get("azure_monitor_workspace_resource_id")
+    if azure_monitor_workspace_resource_id is None or azure_monitor_workspace_resource_id == "":
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/amw/helper.py:29]:
```suggestion
    if not azure_monitor_workspace_resource_id:
```

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/azuremonitorprofile.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+):
+    cloud_name = cmd.cli_ctx.cloud.name
+    if cloud_name.lower() == 'azurechinacloud':
+        raise CLIError("Azure China Cloud is not supported for the Azure Monitor Metrics addon")
+
+    if cloud_name.lower() == "azureusgovernment":
+        grafana_resource_id = raw_parameters.get("grafana_resource_id")
+        if grafana_resource_id is not None:
+            if grafana_resource_id != "":
+                raise InvalidArgumentValueError("Azure US Government cloud does not support Azure Managed Grarfana yet. Please follow this documenation for enabling it via the public cloud : aka.ms/ama-grafana-link-ff")
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/azuremonitorprofile.py:75]:
```suggestion
        if grafana_resource_id:
            raise InvalidArgumentValueError("Azure US Government cloud does not support Azure Managed Grarfana yet. Please follow this documenation for enabling it via the public cloud : aka.ms/ama-grafana-link-ff")
```

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/dc/delete.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        data = json.loads(r.text)
+        dc_object_array = []
+        for item in data['value']:
+            if 'properties' in item and 'dataCollectionRuleId' in item['properties']:
+                dce_id = get_dce_from_dcr(cmd, item['properties']['dataCollectionRuleId'])
+                dc_object_array.append({'name': item['name'], 'dataCollectionRuleId': item['properties']['dataCollectionRuleId'], 'dceId': dce_id})
+        return dc_object_array
+    except CLIError as e:
+        error = e
+        raise CLIError(error)
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/dc/delete.py:44]:
```suggestion
    except CLIError as e:
        raise CLIError(e)
```

File: src/aks-preview/azext_aks_preview/azuremonitormetrics/helper.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    for value in values_array:
+        if value["namespace"].lower() == "microsoft.insights" and value["registrationState"].lower() == "registered":
+            isInsightsRpRegistered = True
+        if value["namespace"].lower() == "microsoft.alertsmanagement" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+        if value["namespace"].lower() == "microsoft.monitor" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+        if value["namespace"].lower() == "microsoft.dashboard" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+    if isInsightsRpRegistered is False:
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/azuremonitormetrics/helper.py:69]:
```suggestion
    if not isInsightsRpRegistered:
```

**Review by yanzhudd — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**